docs/design/resolve-divergence.md

```diff
@@ -0,0 +1,104 @@
+
+1. Starting at the root $R$ of $EvologGraph$.
+1. If $R$ has no successors we exit this loop.
+1. Put the successors of $R$ in one of two sets: those that have no predecessor other than $R$ are added to set $next$, the rest are added to $remaining$.
+1. MergeDivergentTrees: merges the trees of the commits in $next$ with $R$ as base, producing $R'$ (see below). This may result in merge conflicts.
+1. If $R$ is in $DivergentCommits$, add its commit-id to $CommitsToHide$ and remove it from $DivergentCommits$.
+1. Add the intersection of $next$ and $DivergentCommits$ to $CommitsToHide$ and remove them from $DivergentCommits$.
+1. Add $R'$ to $DivergentCommits$.
+1. For each comming $X$ in $remaining$: replace the $R->X$ edge and any edge from $next$ to $X$ with an edge from $R'$ to $X$, while keeping other predecessors of $X$. In other words, $X$ is now a successor of $R'$.
+1. Similarly, for each successor $X$ of any commit in $next$, make $X$ a successor of $R'$, while keeping other predecessors of $X$ (if any).
+1. Set $R$ = $R'$ and go back to step 2.
+
+We repeat the steps above until we exit the loop with a single commit $R$ with the given description and parents. In the happy case the tree in commit $R$ does not have conflicts, but sometimes it will have conflicts. ResolveDivergentChange returns a summary of the changes made, including $R$ and $CommitsToHide$.
+
+The algorithm describe above moves in the "forward" direction. A similar "backwards" algorithm (possibly simpler?) can be implemented. It is hard to say if one is better than the other, it probably depends heavily on the specific situation. It is quite possible we will need heuristics to decide when to give up, for example if conflicts with many sides are produced.
+
+### MergeDivergentTrees
+
+This takes $successors$ (a non-empty set of commits), and $base$ (a single commit). If $successors$ has just one commit, returns that commit. Say $successors$ is $K, L, M$:
+
+* Applies the interdiff between $base$ and $M$ to $L$ to produce $L'$
+* Applies the interdiff between $base$ and $L'$ to $K$ to produce $K'$
+* Returns $K'$
+
+### Rebasing Descendants and Persisting
+
+The last step is to rebase all descendants of $CommitsToHide$ on top of the new commit for $C$ (commit $R$ above), persist the changes and record the operation in the op log. $CommitsToHide$ become hidden, $R$ becomes visible.
+
+### Limiting the set of divergent commits to work on
+
+In some cases the user may want to instruct jj to focus on specific commits in the divergent set. We should consider command-line arg(s) for doing that. It seems this should be fairly straightforward to do. The user provided commit ids must all be visible and all have the same change-id. We simply set $DivergentCommits$ to the user provided value. There may be very rare cases where other commits get picked up by the truncated evolog graph, but that's fine.
+
+### Updating bookmarks
+
+TODO: Write this.
```

Comment:
nit: Missing alternatives (in this case it'd be some alternative algorithms) and or future possibilities section.